File: Composite/python/Composite.py
import abc


class Item(metaclass=abc.ABCMeta):

    def __init__(self, name: str):
        self._item_name: str = name
        self._owner_name: str = None

    def set_owner(self, o: str):
        self._owner_name = o

    @abc.abstractmethod
    def add(self, sub_item: 'Item'):
        pass

    @abc.abstractmethod
    def remove(self, sub_item: 'Item'):
        pass

    @abc.abstractmethod
    def display(self):
        pass


class ClickableItem(Item):

    def __init__(self, name: str):
        super().__init__(name)

    def add(self, sub_item: Item):
        raise Exception('exception')

    def remove(self, sub_item: 'Item'):
        raise Exception('exception')

    def display(self):
        print(self._owner_name + self._item_name)


class DropDownItem(Item):

    def __init__(self, name: str):
        super().__init__(name)
        self.__children = []

    def add(self, sub_item: 'Item'):
        sub_item.set_owner(self._item_name)
        self.__children.append(sub_item)

    def remove(self, sub_item: 'Item'):
        self.__children.remove(sub_item)

    def display(self):
        for item in self.__children:
            if self._owner_name is not None:
                print(self._owner_name, end='')
            item.display()


if __name__ == '__main__':

    file: Item = DropDownItem('file->')

    create: Item = DropDownItem('create->')
    open: Item = DropDownItem('open->')
    exit: Item = ClickableItem('exit')

    file.add(create)
    file.add(open)
    file.add(exit)

    project: Item = ClickableItem('project...')
    repository: Item = ClickableItem('repository...')

    create.add(project)
    create.add(repository)

    solution: Item = ClickableItem('solution...')
    folder: Item = ClickableItem('folder...')

    open.add(solution)
    open.add(folder)

    file.display()
    print()

    file.remove(create)

    file.display()

File: Composite/c++/Composite.cpp
#include <iostream>
#include <string>
#include <vector>
#include <algorithm>

class Item
{
    protected:
        std::string itemName;
        std::string ownerName;

    public:
        Item(std::string name) : itemName(name) {}
        void SetOwner(std::string o);
        virtual void add(Item* subItem) = 0;
        virtual void remove(Item* subItem) = 0;
        virtual void display() = 0;
};

class ClickableItem : public Item
{
    public:
        ClickableItem(std::string name) : Item(name) {}
        void add(Item* subItem) override;
        void remove(Item* subItem) override;
        void display() override;
};

class DropDownItem : public Item
{
    private:
        std::vector<Item*> children;

    public:
        DropDownItem(std::string name) : Item(name) {}
        void add(Item* subItem) override;
        void remove(Item* subItem) override;
        void display() override;
};

int main()
{
    Item* file = new DropDownItem("file->");

    Item* create = new DropDownItem("create->");
    Item* open = new DropDownItem("open->");
    Item* exit = new ClickableItem("exit");

    file->add(create);
    file->add(open);
    file->add(exit);

    Item* project = new ClickableItem("project...");
    Item* repository = new ClickableItem("repository...");

    create->add(project);
    create->add(repository);

    Item* solution = new ClickableItem("solution...");
    Item* folder = new ClickableItem("folder...");

    open->add(solution);
    open->add(folder);

    file->display();
    std::cout << std::endl;

    file->remove(create);

    file->display();

    return 0;
}

void Item::SetOwner(std::string o)
{
    ownerName = o;
}

void ClickableItem::add(Item *subItem)
{
    throw std::exception();
}

void ClickableItem::remove(Item *subItem)
{
    throw std::exception();
}

void ClickableItem::display()
{
    std::cout << itemName << std::endl;
}

void DropDownItem::add(Item *subItem)
{
    subItem->SetOwner(this->itemName);
    children.push_back(subItem);
}

void DropDownItem::remove(Item *subItem)
{
    children.erase(std::remove(children.begin(), children.end(), subItem), children.end());
}

void DropDownItem::display()
{
    for (const auto child : children)
    {
        if (ownerName != "")
            std::cout << ownerName << itemName;
        child->display();
    }
}